File: lang/csharp/src/vpkpp/Format/HOG.cs
using System;

namespace sourcepp.vpkpp.Format;

using OpenPropertyRequest = Func<PackFile, OpenProperty, byte[]>;

using EntryCallback = Action<string, Entry>;

public class HOG : PackFile
{
	protected HOG(nint handle, bool managed = true) : base(handle, managed)
	{
	}

	public new static HOG? Open(string path, EntryCallback? callback = null, OpenPropertyRequest? _ = null)
	{
		var handle = DLL.vpkpp_hog_open(path, callback is not null ? (entryPath, entry) =>
		{
			callback(entryPath, new Entry(entry, false));
		} : null);
		return handle == nint.Zero ? null : new HOG(handle);
	}
}

File: lang/c/src/vpkppc/format/HOG.cpp
#include <vpkppc/format/HOG.h>

#include <sourceppc/Helpers.h>

using namespace sourceppc;
using namespace vpkpp;

const char* VPKPP_HOG_SIGNATURE = HOG_SIGNATURE.data();
const char* VPKPP_HOG_EXTENSION = HOG_EXTENSION.data();

SOURCEPP_API vpkpp_pack_file_handle_t vpkpp_hog_open(const char* path, vpkpp_entry_callback_t callback) {
	SOURCEPP_EARLY_RETURN_VAL(path, nullptr);

	auto packFile = HOG::open(path, callback ? [callback](const std::string& entryPath, const Entry& entry) {
		callback(entryPath.c_str(), const_cast<Entry*>(&entry));
	} : static_cast<PackFile::EntryCallback>(nullptr));
	if (!packFile) {
		return nullptr;
	}
	return packFile.release();
}
